pkg/pulumiyaml/ast/template.go

```diff
@@ -698,6 +699,127 @@ func (d *TemplateDecl) NewDiagnosticWriter(w io.Writer, width uint, color bool)
 	return newDiagnosticWriter(w, fileMap, width, color)
 }
 
+func parseTypeSpec(configDecl *ConfigParamDecl) (schema.TypeSpec, error) {
+	typeSpec := schema.TypeSpec{}
+	if configDecl.Type == nil {
+		return typeSpec, fmt.Errorf("missing type")
+	}
+	switch configDecl.Type.Value {
+	case "string":
+		typeSpec.Type = "string"
+	case "integer":
+		typeSpec.Type = "integer"
+	case "boolean":
+		typeSpec.Type = "boolean"
+	case "array":
+		if configDecl.Items == nil {
+			return typeSpec, fmt.Errorf("missing items")
+		}
+		itemsTypeSpec, err := parseTypeSpec(configDecl.Items)
+		if err != nil {
+			return typeSpec, err
+		}
+		typeSpec.Type = "array"
+		typeSpec.Items = &itemsTypeSpec
+	default:
+		return typeSpec, fmt.Errorf("unknown type: %s", configDecl.Type.Value)
+	}
+	return typeSpec, nil
+}
+
+func (d *TemplateDecl) GenerateSchema() (schema.PackageSpec, error) {
+	description := ""
+	if d.Description != nil {
+		description = d.Description.Value
+	}
+	schemaDef := schema.PackageSpec{
+		Name:        d.Name.Value,
+		Description: description,
+		Language: map[string]schema.RawMessage{
+			"nodejs": schema.RawMessage(`{"respectSchemaVersion": true}`),
+			"python": schema.RawMessage(`{"respectSchemaVersion": true}`),
+			"cshap":  schema.RawMessage(`{"respectSchemaVersion": true}`),
+			"java":   schema.RawMessage(`{"respectSchemaVersion": true}`),
+			"go":     schema.RawMessage(`{"respectSchemaVersion": true}`),
+		},
+	}
+
+	resourcesDef := make(map[string]schema.ResourceSpec)
+	for _, component := range d.Components.Entries {
+		componentType := d.Name.Value + ":index:" + component.Key.Value
+		resourceDef := schema.ResourceSpec{
+			ObjectTypeSpec: schema.ObjectTypeSpec{
+				Properties: map[string]schema.PropertySpec{},
+				Type:       componentType,
+				Required:   []string{},
+			},
+			InputProperties: map[string]schema.PropertySpec{},
+			IsComponent:     true,
+		}
+		if component.Value.Description != nil {
+			resourceDef.Description = component.Value.Description.Value
+		}
+
+		for _, input := range component.Value.Config.Entries {
+			k, v := input.Key.Value, input.Value
+			typeSpec, err := parseTypeSpec(input.Value)
+			if err != nil {
+				return schema.PackageSpec{}, err
+			}
+			def := schemaDefaultValue(v.Default)
+
+			resourceDef.InputProperties[k] = schema.PropertySpec{
+				TypeSpec: typeSpec,
+				Default:  def,
+				DefaultInfo: &schema.DefaultSpec{
+					Environment: []string{k},
+				},
+				Secret: v.Secret != nil && v.Secret.Value,
+			}
+			if def == nil {
+				resourceDef.RequiredInputs = append(resourceDef.RequiredInputs, k)
```

Comment:
Are inputs always required if no default is specified? Can a default be a zero value like `""`? I guess it makes sense if so.

Reply:
Yes, a default can be a zero value like that.  As long as `default` is specified it's fine.